core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -54,10 +58,47 @@ class ZooKeeperClient(connectString: String,
   private val inFlightRequests = new Semaphore(maxInFlightRequests)
   private val stateChangeHandlers = new ConcurrentHashMap[String, StateChangeHandler]().asScala
 
+  private val metricNames = Set[String]()
+
+  // The state map has to be created before creating ZooKeeper since it's needed in the ZooKeeper callback.
+  private val stateToMeterMap = {
+    import KeeperState._
+    val stateToEventTypeMap = Map(
+      Disconnected -> "Disconnects",
+      SyncConnected -> "SyncConnects",
+      AuthFailed -> "AuthFailures",
+      ConnectedReadOnly -> "ReadOnlyConnects",
+      SaslAuthenticated -> "SaslAuthentications",
+      Expired -> "Expires"
+    )
+    stateToEventTypeMap.map { case (state, eventType) =>
+      val name = s"ZooKeeper${eventType}PerSec"
+      metricNames += name
+      state -> newMeter(name, eventType.toLowerCase(Locale.ROOT), TimeUnit.SECONDS)
+    }
+  }
+
   info(s"Initializing a new session to $connectString.")
   @volatile private var zooKeeper = new ZooKeeper(connectString, sessionTimeoutMs, ZooKeeperClientWatcher)
+
+  private val sessionStateGauge =
+    newGauge("SessionState", new Gauge[String] {
+      override def value: String =
+        Option(zooKeeper.getState.toString).getOrElse("DISCONNECTED")
+    })
+
+  metricNames += "SessionState"
+
   waitUntilConnected(connectionTimeoutMs, TimeUnit.MILLISECONDS)
 
+
+  /**
+    * This is added to preserve the original metric name in JMX
+    */
+  override def metricName(name: String, metricTags: scala.collection.Map[String, String]): MetricName = {
+    explicitMetricName("kafka.server", "KafkaHealthcheck", name, metricTags)
```

Comment:
Nit: maybe we can take a `metricGroup` and `metricTypeName` as parameters in the constructor to keep `ZooKeeperClient` generic.

Reply:
Good point. Added.